Big Query -> ClickHouse Equivalent Concepts (#6242)
ClickHouse/clickhouse-docs: docs/cloud/onboard/02_migrate/01_migration_guides/03_bigquery/04_equivalent-concepts.md

## Compute and capacity {#compute-capacity}

| BigQuery | ClickHouse | Notes |
|---|---|---|
| Slot | Replica (whole node); queries can [parallelize](/optimize/query-parallelism) across replicas | A replica is a static compute node in a ClickHouse service. See the callout below. |
| Slot reservation | Vertical and horizontal [autoscaling](/cloud/features/autoscaling/vertical) bounds; [warehouses](/cloud/reference/warehouses) for workload isolation | Where BigQuery reservations guarantee slot capacity, ClickHouse sizes via min/max autoscaling bounds. Setting min = max effectively fixes the size. |
| Quotas | [Workload classes](/operations/workload-scheduling) plus per-query limits | Workload classes cover memory, CPU, concurrency, and I/O scheduling at runtime. Spend caps have no direct primitive; there's no setting that suspends a service on hitting a cost threshold. |

:::note Slot vs replica
Unlike a BigQuery slot, a ClickHouse replica is a static compute
node that serves many concurrent queries; it isn't allocated to a
query from a pool. Replicas are also much coarser-grained: a slot
is closer to a unit of CPU than to a whole node.

> :::

Reviewer: ```suggestion
:::note[Slot vs replica]
Unlike a BigQuery slot, a ClickHouse replica is a static compute
node that serves many concurrent queries; it isn't allocated to a
query from a pool. Replicas are also much coarser-grained: a slot
is closer to a unit of CPU than to a whole node.
:::
```

| Clustering | [`ORDER BY`](/guides/best-practices/sparse-primary-indexes) columns in the table definition | Where BigQuery clustering is best-effort and reorganized in the background, ClickHouse's `ORDER BY` is enforced at insert time and drives the sparse primary index. |
| External tables / BigLake | [`s3`](/sql-reference/table-functions/s3) / [`gcs`](/sql-reference/table-functions/gcs) / [`azureBlobStorage`](/sql-reference/table-functions/azureBlobStorage) table functions for direct file access; [Iceberg engine](/engines/table-engines/integrations/iceberg) for open catalogs | Object storage and open-table formats are read directly through these functions and engines. ClickHouse does not provide a unified-governance layer over external storage. |
| Object tables (SQL access to unstructured files) | [`s3`](/sql-reference/table-functions/s3) / [`gcs`](/sql-reference/table-functions/gcs) table functions over binary formats | Unstructured objects (images, PDFs, audio) are queried directly through these table functions, not via a dedicated table type. |

> | In-SQL ML inference on objects (`ML.GENERATE_TEXT`, `ML.GENERATE_EMBEDDING`) | — | No in-SQL LLM-on-objects equivalent; call hosted models from the application layer. |

Reviewer: ClickHouse does have this now: 

https://clickhouse.com/docs/sql-reference/functions/ai-functions#aiGenerate

https://clickhouse.com/docs/sql-reference/functions/ai-functions#aiEmbed

> | Apache Iceberg | [Iceberg engine](/engines/table-engines/integrations/iceberg) | Reads Iceberg tables stored in S3, Azure, HDFS, or local storage. Writes are not supported. See the engine page for the current list of supported features. |

Reviewer: I think we should try be less specific here WRT read or write support otherwise this will go out of date quickly.

Could we say something like "Wide open-table support" and link to [/use-cases/data-lake/support-matrix](https://clickhouse.com/docs/use-cases/data-lake/support-matrix)

| Table snapshot | Service-level [backup](/cloud/manage/backups/overview) | Granularity differs significantly; see the callout below. |
| Time travel | Point-in-time [backup](/cloud/manage/backups/overview) restore into a new service | No inline historical query; restore is the only way to reach prior state. |
| Authorized views | [View](/sql-reference/statements/create/view) with `SQL SECURITY DEFINER` (runs with the view-owner's privileges) | See [CREATE VIEW](/sql-reference/statements/create/view) for the syntax and the `INVOKER` / `DEFINER` / `NONE` modes. |
| Row-level security | [Row policy](/sql-reference/statements/create/row-policy) — a `WHERE`-style expression evaluated per user | Same role; the policy expression is attached per user (or role) rather than as a BigQuery RLS row access policy resource. |
| Wildcard tables (`_TABLE_SUFFIX`) | [`Merge`](/engines/table-engines/special/merge) table engine (persistent grouping) or [`merge()`](/sql-reference/table-functions/merge) function (inline) | Same idea, different syntax. `Merge` is a persistent table-of-tables; `merge()` is inline without creating one. |

> | Table clone | [`CREATE TABLE ... AS SELECT`](/sql-reference/statements/create/table) copy, or [backup](/cloud/manage/backups/overview) restore into a new service | ClickHouse has no copy-on-write primitive; every copy reads the source data fully. |

Reviewer: ClickHouse supports `CREATE TABLE staging CLONE AS production` syntax, which creates an exact copy of a table without copying the data.

https://clickhouse.com/blog/table-cloning

:::note Backups
ClickHouse Cloud backups are per-service. Restoring
a backup creates a new service; a single table cannot be restored
back into the original service.

> :::

Reviewer: ```suggestion
:::note[Backups]
ClickHouse Cloud backups via the UI are per-service, and create a new service on restore. Finer backup granularity is possible with [`BACKUP`/`RESTORE` SQL commands](s/operations/backup/overview#syntax):::
```

`DELETE`, dbt incremental updates) typically port to engine choice in ClickHouse:
[`ReplacingMergeTree`](/engines/table-engines/mergetree-family/replacingmergetree)
keeps the latest row by sort key, [`CollapsingMergeTree`](/engines/table-engines/mergetree-family/collapsingmergetree)
marks deletes inline, and [`AggregatingMergeTree`](/engines/table-engines/mergetree-family/aggregatingmergetree)
maintains aggregated state. Engine choice is set at table creation and is
non-trivial to change later.
:::

## Query model and performance {#query-model}

Query acceleration in ClickHouse comes from three layers: primary-key ordering (a sparse index over the on-disk sort order), secondary indexes on non-key columns, and materialized views.

| BigQuery | ClickHouse | Notes |
|---|---|---|
| Primary key (advisory) | Primary key — drives the on-disk sort order and the [sparse primary index](/guides/best-practices/sparse-primary-indexes) | Where BigQuery's PK is advisory only, ClickHouse's PK is load-bearing — it determines physical layout and is used to prune granules, avoid re-sorts, and short-circuit `LIMIT`. Neither system enforces uniqueness. |

> | Foreign key (advisory) | Wide tables or [dictionaries](/dictionary) for lookups | ClickHouse doesn't accept foreign-key declarations even as advisory hints. |

Reviewer: ```suggestion
| Foreign key (advisory) | Wide tables or [dictionaries](/dictionary) for lookups | ClickHouse has no notion of foreign keys. This doesn't prohibit joins but means referential integrity is left to the user to manage at an application level. |
```

| Search index | [Full-text index](/engines/table-engines/mergetree-family/textindexes) | Token index over string columns. Operator surface differs: BigQuery's `SEARCH()` maps to the `hasToken` / `hasAllTokens` / `hasAnyTokens` / `hasPhrase` family in ClickHouse; `LIKE` and `match` can also use the index when the pattern is tokenizable (see the index doc for restrictions). |

> | Vector index | [`Array(Float32)`](/sql-reference/data-types/array) plus a [vector ANN index](/engines/table-engines/mergetree-family/annindexes) | ClickHouse has no dedicated vector type. Embeddings are stored as `Array(Float32)` and accelerated with an ANN index for approximate nearest-neighbor lookups. |

Reviewer: Not sure that this is accurate. We have the QBit type: https://clickhouse.com/docs/sql-reference/data-types/qbit - this stores vectors at full precision while letting you choose the fine-grained quantization level at search time. Also have BFloat16: https://clickhouse.com/docs/sql-reference/data-types/float#bfloat16

| Materialized view | [Incremental MV](/materialized-view/incremental-materialized-view) — updates on each insert into a base table | Cost is paid at insert time. BigQuery's auto-refreshed MVs map to this incremental model, not to [refreshable MVs](/materialized-view/refreshable-materialized-view). |
| Scheduled query | [Refreshable MV](/materialized-view/refreshable-materialized-view) for query-driven scheduled work; external orchestrator ([dbt](/integrations/dbt), Airflow) for procedural pipelines | Refreshable MVs run on a cron-style schedule and maintain a result table. |

> | Streaming inserts (`tabledata.insertAll`, Storage Write API) | [Asynchronous inserts](/best-practices/selecting-an-insert-strategy#asynchronous-inserts) for server-side batching of small row-by-row writes; [`Buffer` engine](/engines/table-engines/special/buffer) for in-memory accumulation; [ClickPipes](/integrations/clickpipes) for managed streaming pipelines | The raw API surface is `INSERT` over HTTP or the native protocol, but for the row-by-row semantics of BigQuery streaming inserts, async inserts are the closer fit, batching on the server. For deduplication, ClickHouse exposes block-level dedup on Replicated tables (`insert_deduplicate`, default-on, recent-blocks window) and user-supplied per-insert dedup via `insert_deduplication_token`; for row-level dedup, use `ReplacingMergeTree`. None of these reproduces the Storage Write API's full exactly-once-by-stream-offset model. |

Reviewer: Generally true, but we do support Apache Arrow: https://clickhouse.com/blog/streaming-real-time-visualizations-clickhouse-apache-arrow-perpsective

| VPC Service Controls | [Private connectivity](/cloud/security/connectivity/private-networking) — Private Service Connect (GCP), PrivateLink (AWS, Azure), and IP allowlists for ingress restriction | Restricts ingress; doesn't replicate VPC SC's data-exfiltration boundary. |

## Data sharing {#data-sharing}

| BigQuery | ClickHouse | Notes |
|---|---|---|
| Analytics Hub / data exchanges / listings | Read access to a shared database, or a dedicated service with consumer-specific [row policies](/sql-reference/statements/create/row-policy) | ClickHouse has no in-product data marketplace; sharing uses standard access primitives. |
| Data clean rooms | [Row policies](/sql-reference/statements/create/row-policy) and [authorized views](/sql-reference/statements/create/view) | No managed clean-room product. |

## Operations and ecosystem {#operations}

ClickHouse surfaces operational state through `system.*` tables (queries, sessions, replication, parts, metrics) and the cloud console; managed ingestion is handled by ClickPipes; ML, BI, and notebook workflows are typically handled in external systems that read from ClickHouse.

| BigQuery | ClickHouse | Notes |
|---|---|---|

> | BigQuery ML | External training and serving (notebooks, Spark, Vertex AI, feature stores) reading from ClickHouse; see [AI/ML in Cloud](/cloud/features/ai-ml) for managed-side features | ClickHouse has no in-database ML. Use it as the analytical store and run training elsewhere. |

Reviewer: chDB offers a way to easily integrate ClickHouse into your training pipelines

| Data Transfer Service | [ClickPipes](/integrations/clickpipes) — scheduled and streaming ingestion from SaaS, storage, and OLTP sources | ClickPipes is ClickHouse Cloud's managed connector platform; coverage spans streaming systems, OLTP sources, and object storage. |
| Audit logs (admin activity, data access, system events) | [Cloud audit log](/cloud/security/audit-logging) for org and service admin events; [`system.query_log`](/operations/system-tables/query_log) for data-access activity within the service | BigQuery's three audit streams collapse to two on ClickHouse Cloud: admin activity to the cloud audit log, query/data access to `system.query_log`. System-event telemetry is exposed through other `system.*` tables rather than a dedicated audit stream. |
| Change data capture ingestion | [ClickPipes for Postgres](/integrations/clickpipes/postgres), [MySQL](/integrations/clickpipes/mysql), or Kafka | Managed CDC from OLTP sources. |

> | BigQuery Studio notebooks / BigQuery DataFrames | Jupyter with [`clickhouse-connect`](/integrations/python) or another [client library](/integrations/python) | No in-product notebook environment or pandas-compatible in-DB API; notebook-side libraries cover the same workflow. |

Reviewer: Native hex integration: https://clickhouse.com/integrations/hex

| Data Canvas | — | No drag-and-drop NL canvas. [SQL Console](/integrations/sql-clients/sql-console) covers ad-hoc query authoring; visual data prep happens in an external orchestrator. |

> | Gemini in BigQuery (SQL generation, code completion) | [ClickHouse Agents](/cloud/features/ai-ml/agents) in the cloud console; Ask-AI in the docs | Agents (currently in beta) are conversational: natural-language queries against your data with tool calls and chat workflows. In-SQL LLM functions analogous to `BQ.ML.GENERATE_TEXT` are handled outside SQL today; check the [Agents](/cloud/features/ai-ml/agents) page for the current capability surface. |

Reviewer: We have AI code generation and completion in SQL console via the ClickHouse Assistant: https://clickhouse.com/docs/use-cases/AI_ML/AIChat

I think we can exclude Ask-AI in the docs here, not sure if relevant.

We do have in-SQL AI functions: https://clickhouse.com/docs/sql-reference/functions/ai-functions
